keps/77-dynamically-sized-jobs/README.md

```diff
@@ -0,0 +1,255 @@
+
+## Motivation
+
+Kueue currently lacks support for resizing jobs. When a job is resized, Kueue will recreate the Workload representation of the job, leading to a disruptive suspend and requeue process. This limitation hinders the usability of Kueue for resources like RayCluster that sometimes have autoscaling capabilities enabled.
+
+To properly support RayCluster, Kueue needs to gracefully handle the scale up and scale down of RayCluster nodes. Concretely this means that scaling the resources used by a RayCluster is appropriately reflected in the respective ClusterQueue without needing to suspend the entire RayCluster.
+
+### Goals
+
+- Gracefully handle resize operations for Kueue jobs (i.e. update quota usage without suspend, enqueue scale ups)
+- Autoscaling RayCluster works with Kueue since it has an autoscaler and there is high demand (MVP)s
+
+### Non-Goals
+
+- Vertical scaling of workloads – Kueue will only handle resize operations that scale Pods horizontally 
+- Support resize for other Kueue jobs such as Job, JobSet, etc (future)
+- Resizing of RayJobs
+- Partial Preemption
+
+## Proposal
+
+Update the Job framework reconciler and introduce new controllers to orchestrate dynamic resizing of jobs. We are only interested in horizontal scaling of jobs (e.g. scaling more replicas). At a high level, this will be accomplished by:
+- Creating Workload Slice objects that represent incremental scale up of jobs. This gives us per-replica control of workload admission. Workload Slices will be garbage collected and consolidated with their parent workloads after successful admission.
+- Adding default scheduling gates to control the scheduling of new pods based on their admission status.
+- Dynamically adjust quotas in ClusterQueue based on scaling events.
+
+For the MVP, we will only focus on admission of **RayClusters** with autoscaling enabled.
+
+### User Stories
+
+#### Story 1 - RayCluster w/ autoscaling
+
+1. The user creates a RayCluster with `enableInTreeAutoscaling: true`.
+2. Kueue admits the RayCluster based on the requested resources in the head pod and worker pod. 
+3. Given the deman of the job, the Ray Autoscaler adjusts the replicas field as it adds or removes Pods from the cluster.
+4. Kueue will not suspend and requeue the RayCluster, instead it will dynamically update ClusterQueue usage based on the scaling event. 
+
+### Notes/Constraints/Caveats (Optional)
+
+If Kueue needs to preempt the resized RayCluster, it would preempt it as a whole, regardless of whether the RayCluster was previously scaled up.
+
+## Design Details
+
+### Workload Slices
+To support horizontal scaling of jobs, we will introduce the concept of a "Workload Slice”. A Workload Slice is a Workload object with an owner reference to the original Workload for a job. Workload Slices represent per-replica changes to a job that were not initially accounted for when the job was created.
```

Comment:
> Do the workload slices need to be submitted to the same LocalQueue, or just the same ClusterQueue is enough?

Because LocalQueue is referenced by the top-level RayCluster, it will be submitted to the same LocalQueue

> Do the workload slices need to use identical PodTemplates?

Yes, this is already enforced by KubeRay

> If the two need to be true, what if there is a workload slice that violates the constraints?

From our initial prototyping so far, I think it will be hard to break these constraints. Both the LocalQueue and Pod template are referenced top-level in the RayCluster API.

> 3. Are the workload slices admitted, by Kueue scheduler. to the same resource flavors? (I suppose this is required so that we can aggregate them)

Yes, I think from discussions with Aldo we agreed that all workload slices need to be belong to the same resource flavor

Reply:
> Yes, this is already enforced by KubeRay

Are the Workload slices going to be created by KubeRay?

> From our initial prototyping so far, I think it will be hard to break these constraints. Both the LocalQueue and Pod template are referenced top-level in the RayCluster API.

The constraints probably can be violated if a user creates the Workload slices by hand. Might be worth a note how we handle a mismatch.

Reply:
prebuilt workload feature for context #1575